src/datasets/packaged_modules/base/autofolder.py

```diff
@@ -0,0 +1,398 @@
+import collections
+import itertools
+import os
+from dataclasses import dataclass
+from typing import ClassVar, List, Optional, Tuple
+
+import pyarrow.compute as pc
+import pyarrow.json as paj
+
+import datasets
+
+
+logger = datasets.utils.logging.get_logger(__name__)
+
+
+if datasets.config.PYARROW_VERSION.major >= 7:
+
+    def pa_table_to_pylist(table):
+        return table.to_pylist()
+
+else:
+
+    def pa_table_to_pylist(table):
+        keys = table.column_names
+        values = table.to_pydict().values()
+        return [{k: v for k, v in zip(keys, row_values)} for row_values in zip(*values)]
+
+
+def count_path_segments(path):
+    cnt = 0
+    while True:
+        parts = os.path.split(path)
+        if parts[0] == path:
+            break
+        elif parts[1] == path:
+            break
+        else:
+            path = parts[0]
+            cnt += 1
+    return cnt
+
+
+@dataclass
+class AutoFolderConfig(datasets.BuilderConfig):
+    """BuilderConfig for AutoFolder."""
+
+    _base_feature: ClassVar = (
+        None  # i.e. datasets.Image(), datasets.Audio(), we don't have a type for Feature ("FieldType")
+    )
+    base_feature_name: str = ""
+    label_column: str = "label"
+    features: Optional[datasets.Features] = None
+    drop_labels: bool = None
+    drop_metadata: bool = None
+
+    def __post_init__(self):
+        if not self.base_feature_name:
+            # automatically infer feature name from feature if it's not provided
+            if hasattr(self._base_feature, "_type"):
+                self.base_feature_name = self._base_feature._type.lower()
+            elif hasattr(self._base_feature, "__name__"):
+                self.base_feature_name = self._base_feature.__name__.lower()
+            else:
+                raise AttributeError(
+                    """
+                    `base_feature_name` is not provided and cannot be inferred form the base feature,
+                    check the config's `base_feature`.
+                    """
+                )
```

Comment:
it wouldn't work because neither `Image()` nor `Audio()` doesn't have `__name__` attribute. `__class__.__name__` would work but that way it works for anything (including `None`) which is probably not what we want.

Reply:
i'm not sure i understand correctly, you mean add `BASE_COLUMN_NAME` to the `FolderBuilder` as an abstract attribute and then set their values in `ImageFolder` and `AudioFolder`?

Reply:
[fe80766](https://github.com/huggingface/datasets/pull/4530/commits/fe80766ff7e58ff63d0a953c43536584a3b0ae78)